### DISCUSSION TITLE
Rmarkdown should be a markup language or lines of R in .Rmd files should count towards R programming language

### DISCUSSION THREAD

**Discussion by myoung3** [github-linguist/linguist#5263] [Classification]
Yes, I read https://github.com/github/linguist/issues/4119, but [given that Jupyter notebooks are considered a language ](https://github.com/github/linguist/pull/2746), Rmarkdown should be as well.

**Comment by myoung3** [comment]
Currently github considers Rmd files "prose," but there are situations where this makes for confusing language summary results. [This repo](https://github.com/lindeloev/utility-theory) is written almost entirely in .Rmd, but the language summary describes it as mostly javascript and zero R.

**Comment by lildude** [comment]
RMarkdown _is_ already a language in that Linguist already knows about it and has an entry for it as pointed out in my comment at https://github.com/github/linguist/issues/4119#issuecomment-385316332 I think what you mean is having it considered as `markup` instead of `prose` so is considered when calculating the language statistics by default like Jupyter.

I'm apprehensive about doing this for precisely the reason @Alhadis mentioned:

> It's interesting you mention Jupyter Notebook, because we've had quite a few complaints from users unhappy that their projects were classified as "Jupyter Notebook", simply because of one huge `.ipynb` file (which was mostly generated data, not source code).
>
> I predict if we gave RMarkdown the same treatment, we'd get the same issues reported by users whose R projects became "RMarkdown" overnight.

To add to this apprehension, RMarkdown was [added](https://github.com/github/linguist/pull/831) well over 7 years ago (as markup but then [changed to prose the very next day without a PR](https://github.com/github/linguist/commit/b1bb1a7345150fc182ed8e8175a8cdf7084cd14d)) and this is the first request for this change. Suddenly changing it now _will_ surprise a lot of long time users and not necessarily in a good way.

As such, I think we can keep this issue open for a while to see if it gains any traction with others wanting the same behaviour, but I'm not confident nor sure this is the right thing to do in the long run.

> Currently github considers Rmd files "prose," but there are situations where this makes for confusing language summary results. [This repo](https://github.com/lindeloev/utility-theory) is written almost entirely in .Rmd, but the language summary describes it as mostly javascript and zero R.

As mentioned in https://github.com/github/linguist/issues/4119, you'll need to implement an override if you want your files to show in the stats.

**Comment by myoung3** [comment]
Regarding the concern that counting Rmd lines might distort language percentages strangely, I think this is a reasonable concern. On the one hand:
> Jupyter notebooks have an inflated number of lines of code, since they store a lot of metadata. So it doesn't take many notebooks to "take over" a project.
> (from https://github.com/github/linguist/issues/3316)

The above actually isn't true for Rmd. Jupyter files are stored as json, whereas .Rmd are stored as plain text markdown, so there won't be inflation by metadata.  On the other hand, Rmarkdown documents are often used as instructional pieces so they do tend to be "inflated" by a lot of *actual* prose.  

Another option to consider might be to actually count lines of code within Rmarkdown documents that are explicitly R code chunks, and have that total contribute to the R language count.  This would be more work than just counting lines of .Rmd to its own language (although doable https://www.r-bloggers.com/2018/05/create-code-metrics-with-cloc/), but is probably the safest and most reasonable approach to better reflect programming language use in repos given the concerns that you've raised.

**Comment by lildude** [comment]
> On the other hand, Rmarkdown documents are often used as instructional pieces so they do tend to be "inflated" by a lot of _actual_ prose.

I think you may have hit the nail on the head there as to why this was switched to `prose`. If most files are mostly prose, it makes sense to mark them as such, especially as Linguist doesn't have the ability to do partial file analysis thus...

> Another option to consider might be to actually count lines of code within Rmarkdown documents that are explicitly R code chunks

... is not possible. This is the same issue peeps have had with Jupyter as some want it shown as Python and not Jupyter. 

I think the same explanation and solution I provided at https://github.com/github/linguist/issues/3316#issuecomment-342904191 applies here.

**Comment by sellisd** [comment]
@lildude @myoung3 I did a very rough calculation of how much R-code there is in .Rmd files (as oposed to prose) ([see here](https://gist.github.com/sellisd/b8a3de3b1259484a84d205f16186eee7))

It seems that there is a lot of R code in .Rmd files (about 15% in lines of code) and there is about as much .Rmd files as .R scripts in top R-language repositories in github. These numbers makes me think it is perhaps worth reconsidering if .Rmd should count as R code.
